governance/program/tests/process_execute_instruction.rs

```diff
@@ -0,0 +1,513 @@
+        .await
+        .unwrap();
+
+    let signatory_record_cookie = governance_test
+        .with_signatory(&proposal_cookie, &token_owner_record_cookie)
+        .await
+        .unwrap();
+
+    let proposal_instruction_cookie = governance_test
+        .with_upgrade_program_instruction(
+            &program_governance_cookie,
+            &mut proposal_cookie,
+            &token_owner_record_cookie,
+        )
+        .await
+        .unwrap();
+
+    governance_test
+        .sign_off_proposal(&proposal_cookie, &signatory_record_cookie)
+        .await
+        .unwrap();
+
+    governance_test
+        .with_cast_vote(&proposal_cookie, &token_owner_record_cookie, Vote::Yes)
+        .await
+        .unwrap();
+
+    // Advance slot past hold_up_time
+    let execute_at_slot = 1 + proposal_instruction_cookie.account.hold_up_time + 1;
+
+    governance_test
+        .context
+        .warp_to_slot(execute_at_slot)
+        .unwrap();
+
+    // Ensure we can invoke the governed program before upgrade
+    let governed_program_instruction = Instruction::new_with_bytes(
+        governed_program_cookie.address,
+        &[0],
+        vec![
+            AccountMeta::new(governed_program_cookie.address, false),
+            AccountMeta::new(clock::id(), false),
+            AccountMeta::new(fees::id(), false),
+        ],
+    );
+
+    let err = governance_test
+        .process_transaction(&[governed_program_instruction.clone()], None)
+        .await
+        .err()
+        .unwrap();
+
+    // solana_bpf_rust_upgradable returns CustomError == 42
+    assert_eq!(ProgramError::Custom(42), err);
+
+    // Act
+    governance_test
+        .execute_instruction(&proposal_cookie, &proposal_instruction_cookie)
+        .await
+        .unwrap();
+
+    // Assert
+
+    let proposal_account = governance_test
+        .get_proposal_account(&proposal_cookie.address)
+        .await;
+
+    assert_eq!(1, proposal_account.instructions_executed_count);
+    assert_eq!(ProposalState::Completed, proposal_account.state);
+    assert_eq!(Some(execute_at_slot), proposal_account.closed_at);
+    assert_eq!(Some(execute_at_slot), proposal_account.executing_at);
+
+    let proposal_instruction_account = governance_test
+        .get_proposal_instruction_account(&proposal_instruction_cookie.address)
+        .await;
+
+    assert_eq!(true, proposal_instruction_account.executed);
+
+    // Assert we can invoke the governed program after upgrade
+    governance_test
+        .context
+        .warp_to_slot(execute_at_slot + 10)
+        .unwrap();
+
+    let err = governance_test
+        .process_transaction(&[governed_program_instruction.clone()], None)
+        .await
+        .err()
+        .unwrap();
+
+    // solana_bpf_rust_upgraded returns CustomError == 43
+    assert_eq!(ProgramError::Custom(43), err);
+
+    // --------------------------- !!! Voila  !!! -----------------------------
```

Comment:
Very exciting indeed! So cool to see this working in a test environment :tada:

Reply:
Yeah :) it was actually very good feeling to see this test going green after all the time we spent working on this program.